Matcher in loader

cro opened the pull request:
### What does this PR do?

Creates a new directory `salt/matchers` and migrates the matcher functions into separate .py files in this directory.  These functions are loaded at minion startup, but can be reloaded with the new `saltutil.refresh_matchers` function.

### Tests written?

Matcher tests apply here.

### Commits signed with GPG?

Yes

Ch3LL committed:
Fix macosx grains when swapusage returns comma

twangboy committed:
Fix documentation to denote seconds when setting

twangboy committed:
Use minutes to set

Add ability to use test=True on the state
More documentation improvements
Add log messages

twangboy committed:
Add some more logging

twangboy committed:
Fix unit tests

twangboy committed:
Fix unit.state.test_powercfg

Add test for test=True

twangboy committed:
Add warn_until Fluorine

twangboy committed:
Fix some docs errata

Commit:
Added to consul pillar the ability to nest consul data at a subkey in pillar and to read non-YAML data.

Commit:
rename parse_yaml to expand_keys

Commit:
reorder comment

Commit:
* moved pillar nesting into ext_pillar()
* added test case for pillar key nesting
* added test case for disabling yaml deserialization

Commit:
Fix behaviour of function latest_version in zypper module when multiple packages are passed to function.
Function now properly return dict with empty string as version if latest packages is already installed, and multiple packages are passed to function

asnell committed:
Update manage_file helptext

'source_hash' param should be 'source_sum'

Ch3LL committed:
Add more tests for salt-ssh state.sls_id

Commit:
Adding testcase for this fix.

dwoz committed:
Remove which mock to get the test passing

This test passes in my local kitchen tests when the which mock is
removed. However, I added this mock to try and get the test passing in
jenkins. The original failure that caused me to add it was removed from
jenkins. This will either fix the test or reveal the other error that
needs to be fixed.

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add matchers directory and move default set of matchers to separate .py

cro committed:
Add matchers directory and move default set of matchers to separate .py

dwoz committed:
Fix failing prepend test

The prepend test is failing when jenkins runs kithchen but is passing in
my test environment. Make sure each test starts without the 'etc' path
existing.

cro committed:
Add some more needed imports

cro committed:
Add some more needed imports

cro committed:
Add some more needed imports

cro committed:
Add some more needed imports

cro committed:
Add some more needed imports

Commit:
Don't error on retcode 0 in libcrypto.OPENSSL_init_crypto call

Fixes #46884

garethgreenaway committed:
When using Salt multi-master, if we encouter a salt master that has not accepted the minion key yet we should not exit right away, rather continue on and try the next salt master available in the list.

Ch3LL committed:
import flaky decorator on mac tests

Commit:
Lint: remove extra blank line

cro committed:
Handle this as a classmethod amd be smarter about setting it up.

cro committed:
Handle this as a classmethod amd be smarter about setting it up.

cro committed:
Handle this as a classmethod amd be smarter about setting it up.

cro committed:
Handle this as a classmethod amd be smarter about setting it up.

cro committed:
Handle this as a classmethod amd be smarter about setting it up.

cro committed:
Typo

cro committed:
Clarify setup

cro committed:
Clarify setup

cro committed:
Clarify setup

cro committed:
Clarify setup

cro committed:
Clarify setup

cro committed:
Lint

cro committed:
Lint

cro committed:
Lint

cro committed:
Lint

cro committed:
Lint

cro committed:
Remove unused imports

cro committed:
Remove unused imports

garethgreenaway committed:
If the channel send is sucessful and does not raise an exception, we set ret to True, in case a previous exception from a previous channel send to another master has sent it to False.

garethgreenaway committed:
Adding some quick documention about why we are setting ret=True following the channel.send.

pritambaral committed:
loader: Fix suffix order when importlib is used

On Python versions 2 and 3.0-3.4, imp.get_suffixes() is used to get the suffixes
in an ordered list, which contains '.py' before '.pyc'. On Python versions 3.5
and above, our code to construct such a list using importlib.machinery
inadvertently places '.pyc' before '.py'. This, combined with the effects of
4951b9a60d9b means that things like #46924 happen.

This fixes #46924 by constructing `SUFFIXES` in the same order
imp.get_suffixes() constructs its result.

gtmanfred committed:
only run pylint on files that change

cro committed:
Split matchers out into separate .py files.  Add refresh_matchers, sync_matchers.

cro committed:
Split matchers out into separate .py files.  Add refresh_matchers, sync_matchers.

cro committed:
Split matchers out into separate .py files.  Add refresh_matchers, sync_matchers.

cro committed:
Split matchers out into separate .py files.  Add refresh_matchers, sync_matchers.

cro committed:
Split matchers out into separate .py files.  Add refresh_matchers, sync_matchers.

cro committed:
Refactor matchers to all use similar function signatures.

cro committed:
Refactor matchers to all use similar function signatures.

cro committed:
Refactor matchers to all use similar function signatures.

cro committed:
Refactor matchers to all use similar function signatures.

cro committed:
Refactor matchers to all use similar function signatures.

gtmanfred commented on salt/minion.py:
```diff
@@ -3199,6 +3195,17 @@ class Matcher(object):
+            function_parts = meth.split('.')
+            function_name = function_parts[0]
+            if not getattr(cls, function_name, False):
+                setattr(cls, function_name, matcher_functions[meth])
```
Why do this instead of just calling the matcher_functions[meth] directly?

cro replied:
I think only because I felt like it would require less refactoring since the matcher functions were methods on the Matcher object.

gtmanfred replied:
That is what i guess, this is good enough for a first pass, but in the future we should refactor all of the other places where the matchers are used. :)

isbm replied:
At least in this case `hasattr` can do it.

gtmanfred commented:
This looks great!

Commit:
Adding options to Jenkins pipline builds

Commit:
Adding correct spacing to options section

gtmanfred committed:
Revert "only run pylint on files that change"

dwoz committed:
Sync modules for state tests

Commit:
Adding PY_COLORS=1 for python programs to use ANSI Colors

Commit:
Adding PY_COLORS=1 as PY_COLORS = 1 instead

nicoonoclaste committed:
modules/acme: Set the private key filemode to 0640

Closes #48627

nicoonoclaste committed:
module/acme: Do not exit early when the certificate already exists

Otherwise, acme.cert might fail to cert permissions on the certificate.

Close #48626

AstraLuma committed:
modules/acme: Use file.check_perms ret-morphing powers

As a bonus, the return value now indicates whether permissions where changed.

nicoonoclaste committed:
acme: Make the private key mode configurable

twangboy committed:
Remove vcredist for Py3

AstraLuma committed:
Fixup some schema expectations

twangboy committed:
Use goto instead of if statement

DmitryKuzmenko commented on salt/minion.py:
```diff
@@ -3208,7 +3215,7 @@ def confirm_top(self, match, data, nodegroups=None):
         if not data:
             log.error('Received bad data when setting the match from the top '
-                      'file')
+                        'file')
```
Unrelated change =)

cro replied:
Fixed

DmitryKuzmenko commented:
@cro is it ready for review?

nicoonoclaste committed:
modules/acme: explicitely ignore the `perms` return value

cro commented:
Yes, but I don't know what's up with the tests.  They pass for me.

terminalmage committed:
docker_container.running: Fix regression in test mode

In the process of overhauling docker_container.running to support
network management for 2018.3.0, This line was deleted when it should
not have been. Adding it back to ensure we exit the state at the proper
place when state is run in test mode.

Ch3LL committed:
Add windows ip module integration tests

Ch3LL committed:
Fix service.disabled test for macosx

cachedout commented:
We're close here. Given that we're moving this into a pluggable system, I think we need to add some public-facing documentation about what a `'matcher` is and how to write one.

cro commented:
I can do that.

isbm commented on salt/matchers/pillar_exact_match.py:
```diff
@@ -0,0 +1,30 @@
+    log.debug('pillar target: %s', tgt)
+    if delimiter not in tgt:
+        log.error('Got insufficient arguments for pillar match '
+                  'statement from master')
```
This is an insufficient message either. :wink: Because it doesn't give any hints what to do and even where it comes from, unless you know where it is and know how to read this code. Can we rephrase this and give at least hint how to fix it?

cro replied:
I went through all these and improved the messages.

isbm commented on salt/matchers/pillar_match.py:
```diff
@@ -0,0 +1,29 @@
+    log.debug('pillar target: %s', tgt)
+    if delimiter not in tgt:
+        log.error('Got insufficient arguments for pillar match '
+                  'statement from master')
```
Same here. OK there are more. Could you please look over all those messages and make a better feedback? Because `log.error` is read by admins, that might have absolutely no clue of Salt's internals (and usually they don't).

cro replied:
Likewise

mirceaulinic reviewed (commented):
Wow, that's amazing! Thanks @cro!

isbm commented on salt/modules/saltutil.py:
```diff
@@ -969,6 +1008,24 @@ def refresh_beacons():
+        ret = __salt__['event.fire']({}, 'matchers_refresh')
+    except KeyError:
+        log.error('Event module not available. Matcher refresh failed.')
+        ret = False  # Effectively a no-op, since we can't really return without an event system
```
From API perspective, can we return the same type as `event.fire`?

isbm commented on salt/matchers/grain_pcre_match.py:
```diff
@@ -0,0 +1,28 @@
+    '''
+    Matches a grain based on regex
+    '''
+    log.debug('grains pcre target: %s', tgt)
```
Maybe we should have here a bit more verbose message that mentions matcher explicitly. When lots of debug/trace is happening, then it is easier to grasp this kind of data.

isbm reviewed (changes requested):
@cro looks good to me, except few little nitpicks that would save lots of time for admins and those that aren't deeply familiar with Salt internals. :wink:

mirceaulinic commented on salt/matchers/ipcidr_match.py:
```diff
@@ -0,0 +1,48 @@
+    try:
+        # Target is an address?
+        tgt = ipaddress.ip_address(tgt)
+    except:  # pylint: disable=bare-except
```
+1 I think that would be useful too, indeed.

Ch3LL committed:
reverse assertion order for service disable test

Ch3LL committed:
move the disable call up in try block

bbinet committed:
Add the "traverse" jinja filter

which can be really useful in states to select some leaf pillar data or
fallback with a default value.

bbinet committed:
Add missing profile argument to grafana4.get_user_data

bbinet committed:
Ignore readOnly option when diffing

readOnly value is true only for datasource stored in config (not created
dynamically through the grafana API)
See: https://github.com/grafana/grafana/pull/9659

bbinet committed:
influxdb: fix conflicts in function arg names with client_args

Rename client_args names so that it won't conflict anymore with some
named parameters of influx module function.

For example, the "create_user" function had a "password" argument that
was conflicting with the "password" named argument from client_args.

Commit:
remove windows build cache

cro committed:
Add docs for loadable matchers.

Ch3LL committed:
Use different pub and ret ports for testprogram integration tests

Commit:
DarkGreen Foreground

Commit:
Lesser logging

Ch3LL committed:
Fix mac service.disable tests

garethgreenaway committed:
Updating psutil.disk_partitions to pull in all mounts not just the physical ones.  Check to see if the mount point from the configuration ends with a $ (regular expression end of line) if not we add one in to ensure that a simple / does not end up matching all mount points.  Updating tests accordingly.

Orlando-Richards committed:
Fix for issue #48677 - return True when no changes are to be made with npm.bootstrap with test=true

cro committed:
Add master-side cache matcher

cro committed:
Add compound pillar_exact_match

cro commented:
@isbm I'm in the process of addressing your suggestions
@cachedout Initial docs have been added

There's also a master-side component that the PR now contains the code for, but the docs are not up to date.  I will fix the docs tomorrow.

cro commented:
Oh, also, @isbm, the content of the individual matcher files was copied directly from salt/utils/minions.py and salt/minion.py.  I'm loathe to change that code as it worked before and I don't want to introduce too many variables when getting this all working.

cro committed:
Add details for master-side to the docs.

bbinet committed:
Add support for dry run (test=True) of grafana states

Fixes #43675

bbinet committed:
Use empty dict for 'changes' instead of None

See #47059 #46985

bbinet committed:
Fix RST syntax issue in doc

mirceaulinic committed:
Add Execution module to manipulate JIRA tickets

mirceaulinic committed:
Add autodoc for the modules.jira_mod

Ch3LL committed:
Add windows pkg module integration tests

Ch3LL committed:
Remove unnecessary jinja in curl.sls file

Commit:
removed trailing whitespace for pylint

Commit:
assign index to '_' in fetch_tree return for pylint

cro commented:
OK, I think I have  everything cleaned up.

dgengtek committed:
Feature: ldap authentication with starttls

fix #48655

terminalmage committed:
Add missing release notes mention of file.blockreplace changes

These were added to the docstrings but not to the release notes for the
releases where the change took place.

dwoz committed:
Call sync_all to fix Windows test case

mirceaulinic committed:
Add several new functions to the napalm module

This commit adds the functions required to gate functionality from the new
Fluorine modules [``nxos_api``](https://docs.saltstack.com/en/develop/ref/modules/all/salt.modules.nxos_api.html)
and [``pyeapi``](https://docs.saltstack.com/en/develop/ref/modules/all/salt.modules.arista_pyeapi.html).
NAPALM has all the underlying dependencies for all of these modules, which is
why we can reuse the code from the mentioned modules (plus ``junos``, ``netmiko``
etc.). This brings a lot of flexibility to the user, as without any further
configuration (or very little for the non-native NAPALM drivers) can execute
arbitrary RPC requests or show commands. This is due to the fact that the code
also forwards the already existing configuration details.
In other words, with this changes, starting with release Fluorine, NAPALM users
will be able to execute commands such as:
``salt '*' napalm.pyeapi_run_commands 'show version'`` with no other changes in
their environment.

These functions are important as they open the gates to extend Salt without
depending on the third party NAPALM library, which massively speeds up the
development of new features (particularly when we need features that wouldn't
have their place into the public NAPALM library).

mirceaulinic committed:
Rename napalm.py to napalm_mod.py

isbm commented:
@cro  did you even looked at my comments at all? I am under impression that you didn't even see them.

isbm reviewed (approved):
@cro Ok, so the working on the matcher itself is another PR. Got it.

rallytime commented:
@cro I just noticed that there are some extra commits in here. Can you clean this up/rebase?

Commit:
Update old utils paths to use new paths

Commit:
add salt cloud support for block device tagging or root devices and such

Commit:
add the correct code

cro committed:
Lint

garethgreenaway committed:
Fixing test_diskusage_nomatch and test_diskusage_match_regex

cro commented on salt/matchers/ipcidr_match.py:
```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+'''
+This is the default ipcidr matcher.
```
I did add some better explanations and some pointers to other docs.

cro commented:
@rallytime Done!

Commit:
add missing closing char

Commit:
[develop] Fix beacon test that fails on python 3

Also update test file to use pytest notation instead of unittest

rallytime reviewed (changes requested):
This is very cool :)

Can you add this to the fluorine release notes?

Commit:
[develop] Fix apache tests that fail on python 3

Also update the test file to use pytest syntax instead of unittest

Commit:
fix indenting

Commit:
should make Name tag only default if the user has not defined it in the cloud configuration file

Commit:
fix pylint not in error

cro committed:
Add matcher doc to Fluorine release notes, clean up some minor doc warnings that appear when building the docs.

cro committed:
restore orphan

cro commented:
Release notes updated to include this feature.

Also fixed a couple of other minor doc-build warnings.

Commit:
rework based on lint suggestions

kuetrzi committed:
Update carbon_return.py

fix TypeError: not all arguments converted during string formatting
caused by extra comma accidentally being added to log statement in unicode literals update

dwoz committed:
Skip test when syslog not available

dwoz committed:
Windows compatible cp test

Commit:
Update DOCBANNER with new SaltConf18 image

And remove enterprise ad

Ch3LL committed:
Update Saltstack Logo banner on docs.saltstack.com

Commit:
make recommended changes, update documentation

Commit:
improve document example

mirceaulinic committed:
Revert "Adding the template execution module, as a helper for CLI and other modules"

This reverts commit 1afe33436a34af5e82ad81d141af6e7ce7deeeee.

mirceaulinic committed:
Revert "Adding the template module for autodoc"

This reverts commit e369357ff3f1b2daa4f0ed26f3ad528f301edea6.

mirceaulinic committed:
Get rid of the template mod dependency and simplify the code and usage of the netmiko module

mirceaulinic committed:
Get rid of the template mod dependency and simplify the code and usage of the nxos_api

mirceaulinic committed:
Get rid of the template mod dependency and simplify the code and usage of the pyeapi

Commit:
Add API log rotation on SUSE package

cro committed:
Test fix

cro committed:
Merge branch 'develop' into matcher_in_loader

cro commented:
PR is messy, closing in favor of fresh one.

cro commented:
Cleaner PR is here: #48785
